src/view/writer.js

```diff
@@ -15,582 +15,597 @@ import DocumentFragment from './documentfragment.js';
+ *
+ * Note that {@link engine.view.DocumentFragment DocumentFragment} is thread like a container.
+ *
+ * @see engine.view.AttributeElement
+ * @see engine.view.ContainerElement
+ * @function engine.view.writer.breakAttributes
+ * @param {engine.view.Position} position Position where to break attributes.
+ * @returns {engine.view.Position} New position after breaking the attributes.
+ */
+export function breakAttributes( position ) {
+	return _breakAttributes( position, false );
+}
 
-			return this._breakAttributes( newPosition, forceSplitText );
-		} else
-		// <p>foo<b><u>{}bar</u></b></p>
-		// <p>foo<b>[]<u>bar</u></b></p>
-		// <p>foo{}<b><u>bar</u></b></p>
-		if ( positionOffset === 0 ) {
-			const newPosition = new Position( positionParent.parent, positionParent.getIndex() );
+/**
+ * Uses {@link engine.view.writer.breakAttributes breakAttributes} method to break attributes on
+ * {@link engine.view.Range#start start} and {@link engine.view.Range#end end} positions of
+ * provided {@link engine.view.Range Range}.
+ *
+ * Throws {@link utils.CKEditorError CKEditorError} `view-writer-invalid-range-container` when
+ * {@link engine.view.Range#start start} and {@link engine.view.Range#end end} positions are not placed inside
+ * same parent container.
+ *
+ * Note that {@link engine.view.DocumentFragment DocumentFragment} is thread like a container.
+ *
+ * @see engine.view.writer.breakAttributes
+ * @function engine.view.writer.breakRange
+ * @param {engine.view.Range} range Range which `start` and `end` positions will be used to break attributes.
+ * @returns {engine.view.Range} New range with located at break positions.
+ */
+export function breakRange( range ) {
+	return _breakRange( range );
+}
 
-			return this._breakAttributes( newPosition, forceSplitText );
-		}
-		// <p>foo<b><u>b{}ar</u></b></p>
-		// <p>foo<b><u>b[]ar</u></b></p>
-		// <p>foo<b><u>b</u>[]<u>ar</u></b></p>
-		// <p>foo<b><u>b</u></b>[]<b><u>ar</u></b></p>
-		else {
-			const offsetAfter = positionParent.getIndex() + 1;
+/**
+ * Merges attribute nodes. It also merges text nodes if needed.
+ * Only {@link engine.view.AttributeElement#isSimilar similar} `attribute` nodes can be merged.
+ *
+ * In following examples `<p>` is a container and `<b>` is an attribute node:
+ *
+ *		<p>foo[]bar</p> -> <p>foo{}bar</p>
+ *		<p><b>foo</b>[]<b>bar</b> -> <p><b>foo{}bar</b></b>
+ *		<p><b foo="bar">a</b>[]<b foo="baz">b</b> -> <p><b foo="bar">a</b>[]<b foo="baz">b</b>
+ *
+ * It will also take care about empty attributes when merging:
+ *
+ *		<p><b>[]</b></p> -> <p>[]</p>
+ *		<p><b>foo</b><i>[]</i><b>bar</b></p> -> <p><b>foo{}bar</b></p>
+ *
+ * @see engine.view.AttributeElement
+ * @see engine.view.ContainerElement
+ * @function engine.view.writer.mergeAttributes
+ * @param {engine.view.Position} position Merge position.
+ * @returns {engine.view.Position} Position after merge.
+ */
+export function mergeAttributes( position ) {
```

Comment:
I'd consider renaming this function. I never liked name `AttributeElement` in the first place, because attribute is something else than element and `AttributeElement` can have elements. This causes confusion.

I'd call this method either `mergeAttributeElements`, `mergeAtPosition` or, simply, `merge` (my vote goes here and we could rename `breakAttributes` to `break`). `merge` and `break` would be consistent with `insert`, `move`, etc. Also we do not pass any attributes to the function.